src/environment.py

```diff
@@ -0,0 +1,298 @@
+"""Generate container app environment based on charm configuration."""
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import os
+from urllib.parse import urlparse
+
+# Mattermost's default port, and what we expect the image to use
+CONTAINER_PORT = 8065
+# Default port, enforced via envConfig to prevent operator error
+METRICS_PORT = 8067
+DATABASE_NAME = "mattermost"
+REQUIRED_S3_SETTINGS = ["s3_bucket", "s3_region", "s3_access_key_id", "s3_secret_access_key"]
+REQUIRED_SETTINGS = ["mattermost_image_path"]
+REQUIRED_SSO_SETTINGS = ["license", "site_url"]
+SAML_IDP_CRT = "saml-idp.crt"
+
+
+def _env_for_canonical_defaults(config: dict) -> dict:
+    """Return various Mattermost settings particular to Canonical's deployment.
+
+    These settings may be less generally useful, and so they are controlled here as a unit.
+
+    Args:
+        config: dict of the charm's configuration
+
+    Returns:
+        dict of various Mattermost settings particular to Canonical's deployment.
+    """
+    if not config["use_canonical_defaults"]:
```

Comment:
This is a bit confusing, if I call `_env_for_canonical_defaults` I expect to get the configuration back, it doesn't look like that is the case. Perhaps the test whether the configuration is needed should be done by the caller and this function is responsible for returning the configuration? Although that might also make this function redundant since it seems to just return a constant dictionary

Reply:
Similar comment for the other functions in this file

Reply:
I've decided to just create a constant tuple for canonical's defaults like you proposed.  
I haven't done that for the other functions as they are dynamically created